programs/mmm/src/instructions/cnft/sol_cnft_fulfill_buy.rs

```diff
@@ -0,0 +1,418 @@
+            ctx.accounts.bubblegum_program.key(),
+            args.root,
+            data_hash,
+            args.creator_hash,
+            args.nonce,
+            args.index,
+            None, // signer passed through from ctx
+        )?;
+    }
+
+    // 4. Pool owner as buyer pay royalties to creators
+    let royalty_paid = pay_creator_fees_in_sol_cnft(
+        pool.buyside_creator_royalty_bp,
+        seller_receives,
+        &args.metadata_args,
+        creator_accounts,
+        buyside_sol_escrow_account.to_account_info(),
+        buyside_sol_escrow_account_seeds,
+        system_program.to_account_info(),
+    )?;
+
+    // 5. Seller pay buyer, prevent frontrun by pool config changes
+    let payment_amount = total_price
+        .checked_sub(lp_fee)
+        .ok_or(MMMErrorCode::NumericOverflow)?
+        .checked_sub(taker_fee as u64)
+        .ok_or(MMMErrorCode::NumericOverflow)?
+        .checked_sub(royalty_paid)
+        .ok_or(MMMErrorCode::NumericOverflow)?;
+    if payment_amount < args.min_payment_amount {
+        return Err(MMMErrorCode::InvalidRequestedPrice.into());
+    }
+
+    anchor_lang::solana_program::program::invoke_signed(
+        &anchor_lang::solana_program::system_instruction::transfer(
+            buyside_sol_escrow_account.key,
+            &payer.key,
+            payment_amount,
+        ),
+        &[
+            buyside_sol_escrow_account.to_account_info(),
+            payer.to_account_info(),
+            system_program.to_account_info(),
+        ],
+        buyside_sol_escrow_account_seeds,
+    )?;
+
+    // 6. Pay lp fee
+    if lp_fee > 0 {
+        anchor_lang::solana_program::program::invoke_signed(
+            &anchor_lang::solana_program::system_instruction::transfer(
+                buyside_sol_escrow_account.key,
+                owner.key,
+                lp_fee,
+            ),
+            &[
+                buyside_sol_escrow_account.to_account_info(),
+                owner.to_account_info(),
+                system_program.to_account_info(),
+            ],
+            buyside_sol_escrow_account_seeds,
+        )?;
+    }
+
+    // 7. Pay referral fee
+    if referral_fee > 0 {
+        anchor_lang::solana_program::program::invoke_signed(
+            &anchor_lang::solana_program::system_instruction::transfer(
+                buyside_sol_escrow_account.key,
+                referral.key,
+                referral_fee,
+            ),
+            &[
+                buyside_sol_escrow_account.to_account_info(),
+                referral.to_account_info(),
+                system_program.to_account_info(),
+            ],
+            buyside_sol_escrow_account_seeds,
+        )?;
+    }
+
+    // 8. try close accounts
+    try_close_escrow(
+        &buyside_sol_escrow_account.to_account_info(),
+        pool,
+        system_program,
+        buyside_sol_escrow_account_seeds,
+    )?;
+    try_close_sell_state(sell_state, payer.to_account_info())?;
```

Comment:
not related to your PR but just FYI
if the accounts are actually typed, (e.g. the `Pool` and `SellState`), you can use the anchor close() method from the `AccountsClose` trait
```
fn try_close_sell_state(sell_state: Account<'info, SellState>, payer: AccountInfo<'info>) -> Result<()> {
  sell_state.close(payer)?
}
```

Reply:
that's very nice, we should do a full refactor with all new optimizations